specs/AppNotifications/AppNotificationContentSpec/AppNotificationBuilder-spec.md

```diff
@@ -0,0 +1,668 @@
+
+To generate an AppNotification like this one:
+
+![App Notification With Avatar Image](AppNotificationWithAvatarImage.png)
+
+Developers have to write something like this:
+
+```c++
+int sequenceId {1234};
+std::wstring pathToImage {LR"(path\to\my\image.png)"};
+
+winrt::hstring xmlPayload{
+    L"<toast launch = \"action=AppNotificationClick&amp;sequence=" + to_wstring(sequenceId) + L"\">\
+        <visual>\
+            <binding template = \"ToastGeneric\">\
+                <image placement = \"appLogoOverride\" hint-crop=\"circle\" src = \"" + pathToImage + L"\"/>\
+                <text>App Notification with Avatar Image</text>\
+                <text>This is an example message using XML</text>\
+            </binding>\
+        </visual>\
+        <actions>\
+            <action\
+                content = \"Open App\"\
+                arguments = \"action=OpenApp&amp;sequence=" + to_wstring(sequenceId) + L"\"/>\
+        </actions>\
+    </toast>" };
+
+auto toast{ winrt::AppNotification(xmlPayload) };
+winrt::AppNotificationManager::Default().Show(toast);
+```
+
+# Description
+
+To improve the AppNotification construction experience, the Windows App SDK is introducing
+AppNotificationContent, a builder that constructs simple or rich App Notifications for the
+developer.
+
+```c++
+AppNotificationContent()
+    .AddArgument(L"AppNotificationClick")
+    .AddArgument(L"sequence", 1234)
+    .AddImage(Image(LR"(path\to\my\image.png)")
+        .SetImagePlacement(ImagePlacement::AppLogoOverride)
+        .SetUsesCircleCrop(true))
+    .AddText(Text(L"App Notification with Avatar Image"))
+    .AddText(Text(L"This is an example message"))
+    .AddButton(Button(L"Open App")
+        .AddArgument(L"action", "OpenAppButton")
+        .AddArgument(L"sequence", sequenceId))
+    .Show();
+```
+
+# API Components
+
+**AppNotificationContent** The AppNotificationContent component sets up the xml for a <toast>. This
+is the wrapper content that will include any component adding to the AppNotification UI.
+
+Here is the <toast> schema: <toast launch? = string duration? = "long" | "short" displayTimeStamp? =
+string scenario? = "reminder" | "alarm" | "incomingCall" | "urgent" useButtonStyle? = boolean>
+
+  <!-- Child elements -->
+
+visual, audio?, actions? </toast>
+
+These attributes are abstracted away by the AppNotificationContent component APIs:
+
+-   AddArgument(String key)
+-   AddArgument(String key, T value)
+-   SetDuration(enum Duration)
+-   SetUseButtonStyle(Boolean usesButtonStyle)
+-   SetTimeStamp(String timeStamp)
+-   SetScenario(ScenarioType scenarioType)
+
+All the below UI components are able to be appended through the AppNotificationContent component
+APIs:
+
+-   AddText(Text text)
+-   AddButton(Button button)
+-   AddImage(Image image)
+-   AddAudio(Audio audio)
+-   AddTextBox(TextBox textBox)
+-   AddSelectionMenu(SelectionMenu selectionMenu)
+
+Below is an example usage:
+
+```c#
+AppNotificationContent()
```

Comment:
This is lengthier than the original hard-coded XML example above.  It is also lengthier than the XML it emits.  The XML and builder both provide similar concepts.  However, I'm familiar with the structure of XML.  I'd need guidance about exactly how to specify what I wanted, but at least the structure is familiar.  The builder requires that I provide the same information, but the structure is less familiar to me.  Given that the emitted XML is smaller, I question whether this is "easier" for me.

Reply:
The main advantage, I see, is that the builder can ensure that the produced xml is not only compliant to the schema but can also perform extra validations to ensure the xml will be error free (can validate paths are valid, etc.) and optimized for the OS versions. (there are some subtle difference or compatibility issue when running on down-level OSes )

Reply:
The Windows Toolkit which uses Builder Syntax today has *great* adoption numbers and this is simply an extension of that work in WinAppSDK. @andrewleader is that accurate? So by reasoning, it is evident that a large section of developers prefer Builder Syntax to xml payloads.

Reply:
I'm guessing the builder is popular because you get the tooling you need through intellisense and compiler errors. For XML to be equally good, we'd need an XML editor that provides autocompletion and validation. I don't think length is a good measure of what is better. What really matters is "length of time spent on getting it right".